src/main/java/pl/mimuw/worker/service/WorkerService.java

```diff
@@ -27,18 +29,15 @@ public class WorkerService {
     private final WorkerConfiguration workerConfiguration;
     private final ExecutorService executorService;
     private final PubSubTemplate pubSubTemplate;
+    private final MonitorService monitorService;
 
     private final AtomicInteger currentlyProcessing = new AtomicInteger(0);
     private final Map<String, AcknowledgeablePubsubMessage> currentlyProcessingMessages = new ConcurrentHashMap<>();
 
     @Scheduled(cron = "${worker.pullCron}")
     public void pullAndProcessMonitorTaskMessages() {
         log.info("Pulling messages from task queue, time: {}", currentDate());
-        final var messages = pubSubTemplate.pullAndConvert(
-                workerConfiguration.getSubscriptionId(),
-                workerConfiguration.getMaxTasksPerPod() - currentlyProcessing.get(),
-                true, MonitorTaskMessage.class
-        );
+        final var messages = pullMessages();
         log.info("Pulled {} messages from task queue", messages.size());
         messages.forEach(message -> executorService.submit(() -> processMessage(message)));
```

Comment:
Shouldn't we use `ScheduledExecutorService` there? Because when we do sleep, then the thread we are executing the task on is "dead" and we could utilize it's power to process tasks. Basically we can schedule task execution in the future and it should be more efficient.

Reply:
you are right, didn't know it was a thing. please have a look at the new version